### PR TITLE
Fix tensor^tensor derivative for 0 base entries

### PR DESCRIPTION
**Pull request by agolynski**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32063 Fix scalar^tensor derivative for scalars that are zero
* **#32062 Fix tensor^tensor derivative for 0 base entries**

Differential Revision: [D19394259](https://our.internmc.facebook.com/intern/diff/D19394259)

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 09bf71e4:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=09bf71e402b6407e06061c2c79d8f15ca5dcc560).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4204968) caffe2_onnx_py3_6_clang7_ubuntu16_04_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=83345185) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=83345185)) :snowflake:
<details>
<summary>
<code>Jan 10 20:26:17 test/onnx/test_utility_funs.py::TestUtilityFuns::test_constant_fold_reshape Fatal Python error: Segmentation fault</code>
</summary>

```
Jan 10 20:26:15 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_constant_fold_slice_negative_index PASSED [ 99%] 
Jan 10 20:26:15 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_constant_fold_transpose PASSED [ 99%] 
Jan 10 20:26:15 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_constant_fold_transpose_matmul PASSED [ 99%] 
Jan 10 20:26:15 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_constant_fold_unsqueeze PASSED [ 99%] 
Jan 10 20:26:15 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_error_on_data_parallel PASSED [ 99%] 
Jan 10 20:26:15 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_is_in_onnx_export PASSED [ 99%] 
Jan 10 20:26:15 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_strip_doc_string PASSED [ 99%] 
Jan 10 20:26:15 test/onnx/test_utility_funs.py::TestUtilityFuns_opset11::test_validate_dynamic_axes_invalid_input_output_name PASSED [ 99%] 
Jan 10 20:26:15 test/onnx/test_utility_funs.py::TestUtilityFuns::test_constant_fold_concat PASSED [ 99%] 
Jan 10 20:26:15 test/onnx/test_utility_funs.py::TestUtilityFuns::test_constant_fold_lstm PASSED [ 99%] 
Jan 10 20:26:17 test/onnx/test_utility_funs.py::TestUtilityFuns::test_constant_fold_reshape Fatal Python error: Segmentation fault 
Jan 10 20:26:17  
Jan 10 20:26:17 Current thread 0x00007ff09b614700 (most recent call first): 
Jan 10 20:26:17   File "/usr/local/lib/python3.6/dist-packages/_pytest/_io/saferepr.py", line 43 in repr_instance 
Jan 10 20:26:17   File "/usr/lib/python3.6/reprlib.py", line 65 in repr1 
Jan 10 20:26:17   File "/usr/lib/python3.6/reprlib.py", line 55 in repr 
Jan 10 20:26:17   File "/usr/local/lib/python3.6/dist-packages/_pytest/_io/saferepr.py", line 36 in repr 
Jan 10 20:26:17   File "/usr/local/lib/python3.6/dist-packages/_pytest/_io/saferepr.py", line 67 in saferepr 
Jan 10 20:26:17   File "/usr/local/lib/python3.6/dist-packages/_pytest/_code/code.py", line 655 in repr_args 
Jan 10 20:26:17   File "/usr/local/lib/python3.6/dist-packages/_pytest/_code/code.py", line 736 in repr_traceback_entry 
Jan 10 20:26:17   File "/usr/local/lib/python3.6/dist-packages/_pytest/_code/code.py", line 777 in repr_traceback 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: tools/autograd/templates/Functions.cpp [hunk 1]
```diff
@@ -146,8 +146,19 @@ Tensor pow_backward_self(Tensor grad, const Tensor & self, const Tensor & expone
   return at::where(exponent == 0.0, at::zeros({}, grad.options()), grad * exponent * self.pow(exponent - 1));
 }
 
-Tensor pow_backward_exponent(Tensor grad, const Tensor & self, Tensor result) {
-  return at::where(self == 0, at::zeros({}, grad.options()), grad * result * self.log());
+// Caveats:
+// We define d(a^b)/db at a = 0 and b < 0 to be -inf. This is due to
+// * d(a^b)/db -> -inf for a fixed b as a -> +0
+// * convenience of implementation: no need to have a second at::where statement
```
Comment (albanD) [inline tools/autograd/templates/Functions.cpp]: I don't think this point is needed.
The one above where we define it by continuity is all we need.

File: tools/autograd/templates/Functions.cpp [hunk 2]
```diff
@@ -146,8 +146,19 @@ Tensor pow_backward_self(Tensor grad, const Tensor & self, const Tensor & expone
...
+// Currently, tensorflow defines d(a^b)/db = nan for a = 0 and b < 0.
+//
+// We define d(a^b)/db = 0 for a = 0 and b = 0 by continuity as
+// d(a^b)/db = 0 for a = 0 and b > 0.
```
Comment (albanD) [inline tools/autograd/templates/Functions.cpp]: I think you want `a -> +0` here, not `a = 0`.

**Review by albanD — CHANGES_REQUESTED** [review]
Small update for the comments, otherwise looks good.

File: tools/autograd/templates/Functions.cpp (hunk 1 above)
Comment (agolynski) [reply tools/autograd/templates/Functions.cpp]: Done

File: tools/autograd/templates/Functions.cpp (hunk 2 above)
Comment (agolynski) [reply tools/autograd/templates/Functions.cpp]: I thought we wanted continuity along a = 0 while b -> +0?

**Comment by agolynski** [comment]
All done. PTAL?

**Review by albanD — APPROVED** [review]
LGTM

**Pull request closed without merging**